src/main/java/org/jfrog/bamboo/task/ArtifactoryPublishBuildInfoTask.java

```diff
@@ -54,38 +62,101 @@ protected void initTask(@NotNull TaskContext context) {
         logger = context.getBuildLogger();
         PublishBuildInfoContext publishBuildInfoContext = new PublishBuildInfoContext(context.getConfigurationMap());
         buildParamsOverrideManager = new BuildParamsOverrideManager(customVariableContext);
-        buildInfoHelper = BuildInfoHelper.createDeployBuildInfoHelper(context, context.getBuildContext(), environmentVariableAccessor, publishBuildInfoContext.getArtifactoryServerId(), publishBuildInfoContext.getUsername(), publishBuildInfoContext.getPassword(), buildParamsOverrideManager);
+        BuildContext buildContext = context.getBuildContext();
+        buildInfoHelper = BuildInfoHelper.createDeployBuildInfoHelper(context, buildContext, environmentVariableAccessor, publishBuildInfoContext.getArtifactoryServerId(), publishBuildInfoContext.getUsername(), publishBuildInfoContext.getPassword(), getBuildName(buildContext, publishBuildInfoContext), getBuildNumber(buildContext, publishBuildInfoContext), buildParamsOverrideManager);
     }
 
     @NotNull
     @Override
     public TaskResult runTask(@NotNull TaskContext taskContext) throws TaskException {
-        String buildInfoJson = TaskUtils.getAndDeleteAggregatedBuildInfo(taskContext);
-        Build build = buildInfoHelper.getBuilder(taskContext).build();
+        Map<String, String> customBuildData = taskContext.getBuildContext().getBuildResult().getCustomBuildData();
         ArtifactoryBuildInfoClientBuilder clientBuilder = buildInfoHelper.getClientBuilder(taskContext.getBuildLogger(), log);
         try (ArtifactoryBuildInfoClient client = clientBuilder.build()){
-            if (StringUtils.isNotBlank(buildInfoJson)) {
-                GenericData genericData = BuildInfoExtractorUtils.jsonStringToGeneric(buildInfoJson, GenericData.class);
-                for (Build buildFromContext : genericData.getBuilds()) {
-                    build.append(buildFromContext);
-                }
+            String aggregatedBuildsJson = TaskUtils.getAndDeleteAggregatedBuildInfo(taskContext);
+            Build build = buildInfoHelper.getBuilder(taskContext).build();
+            // Aggregate relevant builds to one build.
+            if (StringUtils.isNotBlank(aggregatedBuildsJson)) {
+                addBuildsToPublish(taskContext, build, aggregatedBuildsJson);
             }
             client.sendBuildInfo(build);
-            taskContext.getBuildContext().getBuildResult().getCustomBuildData().put(BUILD_RESULT_SELECTED_SERVER_PARAM, client.getArtifactoryUrl());
+
+            // Add build details to context.
+            addPublishedBuildDetailsToContext(client.getArtifactoryUrl(), build.getName(), build.getNumber(), taskContext, customBuildData);
```

Comment:
By reading the above code, it is hard to understand the function of the ```addPublishedBuildDetailsToContet``` method vs. the above ```addBuildsToPublish``` method. Let's add further descriptions to make it easier to understand the code flow.

Reply:
Please see my changes in the descriptions of these methods.